DDS-826-reworking_search

dmlond opened the pull request:
implements POST /api/v1/search/files_folders

dmlond committed:
Merge remote-tracking branch 'upstream/develop' into DDS-826-reworking_search

dmlond committed:
Merge branch 'DDS-826-reworking_search' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-826-reworking_search

dmlond committed:
changed 'facets' to 'aggs'
Speling

dmlond committed:
NO NEED TO REBUILD YOUR DOCKER IMAGE
the kaminari gems must be included in the Gemfile before the elasticsearch gems
for pagination support

dmlond committed:
Completely refactored to meet the specifications in api_design/DDS-826-reworking_search
Also supports Kaminari pagination.

dmlond committed:
Merge branch 'DDS-826-reworking_search' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-826-reworking_search

dmlond committed:
refactored for new ElasticsearchResponse expectations
 - results and aggs are attributes
 - aggs are filtered from the Serialized output if not supplied
   in the requested query definition

dmlond committed:
reverted to previous version to retaion deprecated functionality for the /search
endpoint.
These objects will be renamed 'DeprecatedElasticsearchResponse' in a subsequent
commit to ease cleanup when we finally remove the endpoint.

dmlond committed:
Deprecating underlying models for /search
These will be removed in a future version

dmlond committed:
FolderFilesResponse was committed as a new ElasticsearchResponse in a previous commit.
The object and serializer will service the /search/folders_files endpoint

dmlond committed:
modified verification error responses to be more generic, consistent

dmlond committed:
uses a better method to tell if an image exists

dmlond committed:
request parameters shared_context no longer sets up a 'payload'

added 'a client error' as a general method to test when a request
  returns the standard 4XX error

added 'a PUT request'  and 'a POST request' methods which are similar
to 'a GET request'

dmlond committed:
These refactored to define a payload since it is no longer defined in
the request parameters context.

uses 'a GET request', 'a POST request', and 'a PUT request' more consistently.

some request tests in the uploads_api_spec cannot use 'a PUT request' for
some reason.

dmlond committed:
implements POST /api/v1/search/folders_files according to DDS-826-reworking_search specifications

dmlond committed:
unsets RECREATE_SEARCH_MAPPINGS if it is set after search mappings rebuilt

dmann commented on app/api/dds/v1/search_api.rb:
```diff
@@ -76,12 +78,81 @@ class SearchAPI < Grape::API
+          optional 'project.id', type: Array, desc: 'Limit search context to list of projects; requires a set of comma-delimited ids like so: {"project.id": ["345e...", "2e45..."]}. Filters out projects to which the user is not allowed access. If not specified, defaults to projects in which the current user has view permission.'
+        end
+        optional :aggs, type: Array, desc: 'List of aggs (aggregates) that should be computed.' do
+          optional :field, type: String, desc: 'The field for which you want to compute agg; valid options are project.name and tags.label. Required if one or more aggs are submitted.'
```
What does grape do when a `requires` is set inside of a `optional` parameter? If it works as expected, maybe we could get away from having "Required if one or more aggs are submitted." in the parameter descriptions.

dmlond replied:
I tried this. It makes it difficult to provide the expected error payload for missing parameters, because it halts processing and returns a response before our code gets the chance to rescue from the exception.

dmann replied:
Makes sense, thanks for the explanation.

dmann commented on app/api/dds/v1/search_api.rb:
```diff
@@ -76,12 +78,81 @@ class SearchAPI < Grape::API
+        search_params[:filters] ||= []
+        search_params[:filters].each do |f|
+          if f.has_key? 'project.id'
+            filtered = policy_scope(Project).where(is_deleted: false, id: f['project.id']).all.to_a.map{|p| p.id}
```
The policy_scope calls are fairly complex and expensive. Since a non-specific error is raised as a result of this, maybe we could refactor this to collect the 'project.id' filters in an array and then flatten that and pass it to the policy scope.

dmlond replied:
There should only ever be at most one 'project.id', or 'kind' entry in the search_params[:filters] array. This will not call the policy_scope more than once per request. I could throw an exception for this in the FoldersFilesResponse.filter or validate_filters method to be sure.

dmlond replied:
I am asking Matt to make sure this is what he intended, and suggesting that filters and post_filters could be a Hash instead of Array.

dmlond replied:
just got confirmation from Matt. He wants to accept an array of filters, so that we can extend our filters in the future to support objects with different schemas, and possible different predicates. BUT he is in agreement that we should only allow one 'project.id' or 'kind' object to be passed in. I am adding this to FolderFilesResponse and testing it.

dmann commented on app/api/dds/v1/search_api.rb:
```diff
@@ -76,12 +78,81 @@ class SearchAPI < Grape::API
+          .query(search_params[:query_string])
+          .aggregate(search_params[:aggs])
+          .post_filter(search_params[:post_filters])
+          .search
```
This message chaining looks really clean. A++  😃

dmann commented on app/models/folder_files_response.rb:
```diff
@@ -0,0 +1,238 @@
+    if @aggs
+      query[:aggs] = {}
+      @aggs.each do |agg|
+        size = agg[:size] || 20
```
Would it be overkill to create a constant to hold the default agg size?

dmlond replied:
I can do that

dmann commented on circle.yml:
```diff
@@ -46,6 +46,7 @@ deployment:
     - heroku run rake --exit-code db:seed --app 'dukeds-dev'
     - heroku run --exit-code rake db:data:migrate --app 'dukeds-dev'
     - heroku run --exit-code rake elasticsearch:index:rebuild --app 'dukeds-dev'
+    - heroku run config:unset RECREATE_SEARCH_MAPPINGS
```
Nice! No more forgetting to unset that config var!

dmann reviewed (changes requested):
Looks good. I really like how the FolderFilesResponse model came out, very clean! Just a few minor changes/refactoring, but that's to be expected with a PR with over 2k lines of new code.

dmlond committed:
added FolderFilesResponse.default_agg_size

dmlond committed:
Modified the api design:
  - only one 'kind' or 'project.id' filter object will be accepted in a single request
  - only one 'project.name' or 'tags.label' post_filter object will be accepted in a single request

FolderFilesResponseSpec expects FolderFilesResponse to meet new filter and post_filter
requirements

FolderFilesResponse passes spec

dmann commented on spec/models/folder_files_response_spec.rb:
```diff
@@ -239,6 +261,24 @@
+          [
+            all_projects,
+            {"kind" => ["dds-folder"]},
+            {"kind" => ["dds-fille"]}
```
Is this a typo? Should this be "dds-file"?

dmlond replied:
yes, I will fix it :)

dmlond committed:
speling

dmlond committed:
use default_agg_size when building the query
